File: 78. Subsets.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        res.push_back({});
        for (int k = 1; k <= nums.size(); k++)
            combination(nums, {}, 0, k);
        return res;
    }
    
    void combination(vector<int>& nums, vector<int> path, int start, int k)
    {
        if (path.size() == k)
            res.push_back(path);
        for (int i = start; i < nums.size(); i++)
        {
            path.push_back(nums[i]);
            combination(nums, path, i + 1, k);
            path.pop_back();
        }
    }
};

File: 78. Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = [[]]
        for k in xrange(1, len(nums)+1):
            self.combination(nums, [], 0, k)
        return self.res
    
    def combination(self, nums, path, start, k):
        if len(path) == k:
            self.res += [path[:]]
            return
        for i in xrange(start, len(nums)):
            self.combination(nums, path + [nums[i]], i+1, k)
